src/Microsoft.FeatureManagement/Targeting/ContextualTargetingFeatureVariantAssigner.cs

```diff
@@ -67,69 +67,128 @@ public ValueTask<FeatureVariant> AssignVariantAsync(FeatureVariantAssignmentCont
+                    // Valid to omit audience for default variant
+                    continue;
+                }
+
+                if (!TargetingEvaluator.TryValidateSettings(targetingSettings, out string paramName, out string reason))
+                {
+                    throw new ArgumentException(reason, paramName);
+                }
+
+                //
+                // Check if the user is being targeted directly
+                if (targetingSettings.Audience.Users != null &&
+                    TargetingEvaluator.IsTargeted(
+                        targetingContext,
+                        targetingSettings.Audience.Users,
+                        _options.IgnoreCase))
+                {
+                    return new ValueTask<FeatureVariant>(v);
+                }
+            }
 
             var cumulativeGroups = new Dictionary<string, double>(
                 _options.IgnoreCase ? StringComparer.OrdinalIgnoreCase :
                                       StringComparer.Ordinal);
 
+            //
+            // Check Groups
             foreach (FeatureVariant v in featureDefinition.Variants)
             {
-                TargetingFilterSettings targetingSettings = v.AssignmentParameters.Get<TargetingFilterSettings>();
+                TargetingFilterSettings targetingSettings = lookup[v];
 
-                if (targetingSettings == null)
+                if (targetingSettings == null ||
+                    targetingSettings.Audience.Groups == null)
                 {
-                    if (v.Default)
-                    {
-                        //
-                        // Valid to omit audience for default variant
-                        continue;
-                    }
+                    continue;
                 }
 
-                if (!TargetingEvaluator.TryValidateSettings(targetingSettings, out string paramName, out string reason))
+                AccumulateGroups(targetingSettings.Audience, cumulativeGroups);
+
+                if (TargetingEvaluator.IsTargeted(
+                    targetingContext,
+                    targetingSettings.Audience.Groups,
+                    _options.IgnoreCase,
+                    featureDefinition.Name))
                 {
-                    throw new ArgumentException(reason, paramName);
+                    return new ValueTask<FeatureVariant>(v);
                 }
+            }
+
+            double cumulativePercentage = 0;
 
-                AccumulateAudience(targetingSettings.Audience, cumulativeGroups, ref cumulativePercentage);
+            //
+            // Check default rollout percentage
+            foreach (FeatureVariant v in featureDefinition.Variants)
+            {
+                TargetingFilterSettings targetingSettings = lookup[v];
 
-                if (TargetingEvaluator.IsTargeted(targetingSettings, targetingContext, _options.IgnoreCase, featureDefinition.Name))
+                if (targetingSettings == null)
```

Comment:
>if (targetingSettings == null) [](http://example.com/codeflow?start=16&length=30)

Should  we skip those variants that do not have any DefaultRolloutPercentage?

Reply:
I guess you mean to special case if `DefaultRolloutPercentage` is 0 ? We could, but the normal logic handles it fine.